README.md
# pokemongo-bot

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-07-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-07-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/pokemongo-bot/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/pokemongo-bot/compare/0.1.0...0.1.1


test/pokemongo_bot/core_test.clj
(ns pokemongo-bot.core-test
  (:require [clojure.test :refer :all]
            [pokemongo-bot.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/pokemongo_bot/api.clj
(ns pokemongo-bot.api
  (:use [clojure.core.async :only [go go-loop chan timeout <! <!! >! >!! close!]])
  (:require
   [pokemongo-bot.authenticate :as auth]
   [pokemongo-bot.coordinate   :as coord]
   [pokemongo-bot.pokemon      :as pokemon]
   [pokemongo-bot.item         :as item])
  (:import
   [okhttp3 OkHttpClient]
   [com.pokegoapi.api PokemonGo]))

(def pokemon-client (ref nil))

(defn client [account]
  (let [http-client (OkHttpClient.)
        api-client  (PokemonGo. (auth/login account http-client) http-client)]
    (dosync (ref-set pokemon-client api-client))
    api-client))

(defn location [point]
  (let [latlong (coord/coord point)]
    (-> @pokemon-client (.setLocation (:latitude  latlong) (:longitude latlong) 1))
    latlong))

(defn current-location []
  (coord/coord @pokemon-client))

(defn inventories []
  (.getInventories @pokemon-client))

(defn inventory-items []
  (try (->> (inventories) .getItemBag .getItems)
       (catch com.pokegoapi.exceptions.AsyncPokemonGoException pge
         (println "アイテムバッグを参照する際の同期処理エラー"))))

(defn remove-item [item count]
  (try (-> (inventories) .getItemBag (.removeItem (.getItemId item) count))
       (catch com.pokegoapi.exceptions.AsyncPokemonGoException pge
         (println "アイテムを捨てる際の同期処理エラー"))))

(defn update-inventory []
  (try
    (-> (inventories) .updateInventories)
    (catch com.pokegoapi.exceptions.AsyncPokemonGoException pge
      (println "インベントリを更新する際の同期処理エラー"))
    (catch java.util.ConcurrentModificationException cme
      (println "インベントリを更新する際に同期処理をしようとしてエラー"))))

(defn transfer-pokemon [pk]
  (try
    (.transferPokemon pk)
    (catch com.pokegoapi.exceptions.AsyncPokemonGoException e
      (println "ポケモンをアメに替えようとしたら同期処理エラー"))))

(defn evolve-pokemon [pk]
  (try
    (.evolve pk)
    (catch com.pokegoapi.exceptions.AsyncPokemonGoException e
      (println "ポケモンを進化させようとしたら同期処理エラー"))))


(defn pokeballs []
  (->> (inventory-items)
       (map item/status)
       (filter #(some (fn [_] (= _ (-> % :item-id :number))) [1 2 3 4]))))

(defn bestball []
  (let [balls (pokeballs)]
    (cond (->> balls (some #(and (= 2  (-> % :item-id :number)) (< 20 (-> % :count)))))
          com.pokegoapi.api.inventory.Pokeball/GREATBALL
          (->> balls (some #(and (= 3  (-> % :item-id :number)) (< 20 (-> % :count)))))
          com.pokegoapi.api.inventory.Pokeball/ULTRABALL)))

(defn catch-pokemon [cpk]
  (let [pokemon-name (-> cpk pokemon/pokemon-id :jname)]
    (print (str pokemon-name "を発見した！"))
    (Thread/sleep 5000)
    (let [encount (try (if (.encounterPokemon cpk) true false)
                     (catch com.pokegoapi.exceptions.AsyncPokemonGoException pge
                       (println "ポケモンとエンカウントしようとしたら同期処理エラー")))]
    (Thread/sleep 10000)
    (try (let [select-ball (bestball)
               result (if select-ball
                        (.catchPokemon cpk select-ball)
                        (.catchPokemon cpk))]
           (println (str pokemon-name (pokemon/result-format result))))
         (catch java.lang.NullPointerException e
           (println "ポケモンを捕まえようとしたら結果が空でパース出来ずエラー"))
         (catch com.pokegoapi.exceptions.AsyncPokemonGoException pge
           (println "ポケモンを捕まえようとしたら同期処理エラー"))))))

(defn pokemon-map []
  (-> @pokemon-client .getMap))

(defn pokemon-spots []
  (try (-> (pokemon-map) .getMapObjects)
       (catch com.pokegoapi.exceptions.AsyncPokemonGoException pge
         (println "マップ情報を取得しようとして同期処理エラー"))))

(defn catchable-pokemons []
  (try (-> (pokemon-map) .getCatchablePokemon)
       (catch com.pokegoapi.exceptions.AsyncPokemonGoException pge
         (println "捕獲可能なポケモンを取得しようとして同期処理エラー"))))

(defn search-pokestops []
  (try (or (-> (pokemon-spots) .getPokestops) [])
       (catch java.lang.NullPointerException e [])))

(defn lootable-pokestops []
  (->> (search-pokestops)
       (filter #(.canLoot %))))

(defn loot-pokestop [pokestop]
  (try
    (.loot pokestop)
    (catch com.pokegoapi.exceptions.AsyncPokemonGoException e
      (println "ポケストップから戦利品を取得しようとして同期処理エラー"))))


(defn start [account coord]
  (client account)
  (location coord)
  (go-loop []
    (<! (timeout (* 1800 1000)))
    (let [current (current-location)]
      (client account)
      (location current))
    (recur)))


src/pokemongo_bot/core.clj
(ns pokemongo-bot.core
  (:gen-class)
  (:import
   [com.pokegoapi.util Log Log$Level])
  (:require
   [pokemongo-bot.authenticate :as auth]
   [pokemongo-bot.coordinate   :as coord]
   [pokemongo-bot.pokemon      :as pokemon]
   [pokemongo-bot.item         :as item]
   [pokemongo-bot.api          :as api]))

(Log/setLevel Log$Level/INFO)

(def item-limits
  {1 300  ;; ITEM_POKE_BALL
   2 150  ;; ITEM_GREAT_BALL
   3 150  ;; ITEM_ULTRA_BALL
   101 0  ;; ITEM_POTION
   102 0  ;; ITEM_SUPER_POTION
   103 0  ;; ITEM_HYPER_POTION
   104 30 ;; ITEM_MAX_POTION
   201 15 ;; ITEM_REVIVE
   202 15 ;; ITEM_REVIVE
   })

(defn remove-item-set [i]
  (let [item   (item/status i)
        number (-> item :item-id :number)
        limit  (item-limits number)]
    (if (and limit (> (item :count) limit))
      (list i (- (item :count) limit)))))

(defn remove-items []
  (->> (api/inventory-items)
       (map remove-item-set)
       (filter identity)
       (map #(apply api/remove-item %))
       doall))

(defn release-pokemon [pk]
  (if (and (some #(= % (:number (pokemon/pokemon-id pk))) pokemon/evolves)
           (< 100 (-> (api/inventories)
                      .getCandyjar
                      (.getCandies (.getPokemonFamily pk)))))
    (do (Thread/sleep 3000)
        (api/evolve-pokemon pk))
    (do (Thread/sleep 3000)
        (api/transfer-pokemon pk))))

(defn release-pokemons [attack defense stamina]
  (if (api/inventories)
    (->> (-> (api/inventories) .getPokebank .getPokemons)
         (filter #(let [status (pokemon/status %)]
                    (if (< 1500 (:cp status))
                      false
                      (or (> attack  (:individual-attack  status))
                          (> defense (:individual-defense status))
                          (> stamina (:individual-stamina status))))))
         (map release-pokemon)
         doall)))

(defn distance-with-center [center current spot]
  (+ (-> (coord/distance current spot) (* 15))
     (-> (coord/distance center spot))))

(defn next-spot [center routes]
  (let [current (api/current-location)
        spots (->> (api/search-pokestops)
                   (filter #(not (some (fn [_] (= _ (coord/coord %))) routes)))
                   (filter #(= 0 (.getCooldownCompleteTimestampMs %)))
                   (sort-by #(distance-with-center center current %)))
        best (first spots)]
    (if best
      (if (= current (coord/coord best))
        (second spots)
        best))))

(defn enough-ball? []
  (->> (api/inventory-items)
       (map item/status)
       (some #(or (and (= 1 (-> % :item-id :number)) (< 20 (-> % :count)))
                  (and (= 2 (-> % :item-id :number)) (< 20 (-> % :count)))
                  (and (= 3 (-> % :item-id :number)) (< 20 (-> % :count)))))))

(defn play-game []
  (if (enough-ball?)
    (->> (api/catchable-pokemons)
         (map api/catch-pokemon)
         doall))
  (->> (api/lootable-pokestops)
       (map api/loot-pokestop)
       doall))

(defn walking [spot cooltime]
  (let [current (api/current-location)]
    (->> (coord/jogging-route current spot cooltime)
         (map #(do (Thread/sleep (* cooltime 1000))
                   (api/location %)
                   (play-game)))
         doall)))

(defn spot-pool [x coll]
  (cons (coord/coord x) (take 4 coll)))

(defn walk-around [center]
  (loop [routes []]
    (api/update-inventory)
    (release-pokemons 13 13 13)
    (remove-items)
    (let [next (next-spot center routes)]
      (if next
        (do (println (str "次のポケストップ : " (coord/coord-format next)))
            (println (str "次のポケストップまでの距離 : "
                          (coord/distance (api/current-location) next)
                          " メートル"))
            (walking next 1)
            (api/loot-pokestop next)
            (println "ポケストップに到着")
            (recur (spot-pool next routes)))
        (do
          (Thread/sleep 20000)
          (println (str "ポケストップが見つからない"))
          (recur routes))))))

(defn walking-start [account center]
  (api/start account center)
  (walk-around center))

(defn -main [& args]
  (let [email    (-> (System/console) (.readLine " Gmail : " nil) String/valueOf)
        password (-> (System/console) (.readPassword " Password : " nil) String/valueOf)
        latlong  (do (println " - Shibuya109           : 35.65955,139.699068")
                     (println " - 東京ディズニーランド : 35.6328964,139.8803943")
                     (println " - 東京ディズニーシー   : 35.6267108 139.8850779")
                     (-> (System/console) (.readLine " Coord : " nil) String/valueOf))]
    (walking-start
     (auth/->GoogleAccount email password)
     (apply coord/->Coord (map read-string (clojure.string/split latlong #",")))
     )))

;; (coord/->Coord 35.6328964 139.8803943) ;; TDL
;; (coord/->Coord 35.65955 139.699068)    ;; Shibuya 109
;; (coord/->Coord 35.6267108 139.8850779) ;; TDS

;; {:number 1,   :name "ITEM_POKE_BALL"}
;; {:number 2,   :name "ITEM_GREAT_BALL"}
;; {:number 3,   :name "ITEM_ULTRA_BALL"}
;; {:number 101, :name "ITEM_POTION"}
;; {:number 102, :name "ITEM_SUPER_POTION"}
;; {:number 103, :name "ITEM_HYPER_POTION"}
;; {:number 104, :name "ITEM_MAX_POTION"}
;; {:number 201, :name "ITEM_REVIVE"}
;; {:number 301, :name "ITEM_LUCKY_EGG"}
;; {:number 401, :name "ITEM_INCENSE_ORDINARY"}
;; {:number 501, :name "ITEM_TROY_DISK"}
;; {:number 701, :name "ITEM_RAZZ_BERRY"}
;; {:number 901, :name "ITEM_INCUBATOR_BASIC_UNLIMITED"}
;; {:number 902, :name "ITEM_INCUBATOR_BASIC"}



src/pokemongo_bot/coordinate.clj
(ns pokemongo-bot.coordinate
  (:import
   [javax.measure.unit SI NonSI]
   [org.jscience.geography.coordinates.crs ReferenceEllipsoid]
   [com.pokegoapi.api PokemonGo]
   [POGOProtos.Map SpawnPointOuterClass$SpawnPoint]
   [com.pokegoapi.api.map.fort Pokestop]))


(defrecord Coord [latitude longitude])

(defrecord UTM [easting northing])

(defn utm [latlong]
  (let [jutm
        (org.jscience.geography.coordinates.UTM/latLongToUtm
         (org.jscience.geography.coordinates.LatLong/valueOf
          (:latitude  latlong)
          (:longitude latlong)
          NonSI/DEGREE_ANGLE)
         ReferenceEllipsoid/WGS84)
        position (.getPosition jutm)]
    (UTM. (-> position (.eastingValue SI/METER))
          (-> position (.northingValue SI/METER)))))


(defprotocol ICoord
  (coord [p])
  (distance [p1 p2])
  (coord-format [p]))

(extend-protocol ICoord
  Coord
  (coord [c] c)
  (distance [c1 c2]
    (let [u1 (utm c1) u2 (utm (coord c2))]
      (Math/sqrt
       (+ (-> (- (:easting  u1) (:easting  u2)) (Math/pow 2))
          (-> (- (:northing u1) (:northing u2)) (Math/pow 2))))))
  (coord-format [c] (str (:latitude c) "," (:longitude c)))

  PokemonGo
  (coord [pg] (coord (Coord. (.getLatitude pg) (.getLongitude pg))))
  (distance [pg p1] (distance (coord pg) (coord p1)))
  (coord-format [pg] (coord-format (coord pg)))

  SpawnPointOuterClass$SpawnPoint
  (coord [sp] (coord (Coord. (.getLatitude sp) (.getLongitude sp))))
  (distance [sp p1] (distance (coord sp) (coord p1)))
  (coord-format [sp] (coord-format (coord sp)))

  Pokestop
  (coord [pks] (coord (Coord. (-> pks .getFortData .getLatitude)
                              (-> pks .getFortData .getLongitude))))
  (distance [pks p1] (distance (coord pks) (coord p1)))
  (coord-format [sp] (coord-format (coord sp))))

(defn route [start end km-h cooltime]
  (let [m-s (-> km-h (* 1000) (/ 3600))
        start (coord start)
        end   (coord end)
        times (Math/ceil (/ (distance start end) (* m-s cooltime)))
        move-lat  (-> (:latitude  end) (- (:latitude  start)) (/ times))
        move-long (-> (:longitude end) (- (:longitude start)) (/ times))]
    (conj
     (vec (map #(Coord. %1 %2)
               (range (:latitude  start) (:latitude  end) move-lat)
               (range (:longitude start) (:longitude end) move-long)))
     end)))

(defn walking-route [start end cooltime]
  (route start end 5 cooltime))

(defn jogging-route [start end cooltime]
  (route start end 10 cooltime))




src/pokemongo_bot/pokemon.clj
(ns pokemongo-bot.pokemon
  (:import
   [com.pokegoapi.api.map.pokemon CatchablePokemon]
   [com.pokegoapi.api.pokemon Pokemon]
   [POGOProtos.Enums PokemonIdOuterClass$PokemonId]
   [com.pokegoapi.api.map.pokemon CatchResult]
   [POGOProtos.Networking.Responses
    CatchPokemonResponseOuterClass$CatchPokemonResponse$CatchStatus]))

(def ^:dynamic *pokemons*
  {1 "フシギダネ"
   2 "フシギソウ"
   3 "フシギバナ"
   4 "ヒトカゲ"
   5 "リザード"
   6 "リザードン"
   7 "ゼニガメ"
   8 "カメール"
   9 "カメックス"
   10 "キャタピー"
   11 "トランセル"
   12 "バタフリー"
   13 "ビードル"
   14 "コクーン"
   15 "スピアー"
   16 "ポッポ"
   17 "ピジョン"
   18 "ピジョット"
   19 "コラッタ"
   20 "ラッタ"
   21 "オニスズメ"
   22 "オニドリル"
   23 "アーボ"
   24 "アーボック"
   25 "ピカチュウ"
   26 "ライチュウ"
   27 "サンド"
   28 "サンドパン"
   29 "ニドラン♀"
   30 "ニドリーナ"
   31 "ニドクイン"
   32 "ニドラン♂"
   33 "ニドリーノ"
   34 "ニドキング"
   35 "ピッピ"
   36 "ピクシー"
   37 "ロコン"
   38 "キュウコン"
   39 "プリン"
   40 "プクリン"
   41 "ズバット"
   42 "ゴルバット"
   43 "ナゾノクサ"
   44 "クサイハナ"
   45 "ラフレシア"
   46 "パラス"
   47 "パラセクト"
   48 "コンパン"
   49 "モルフォン"
   50 "ディグダ"
   51 "ダグトリオ"
   52 "ニャース"
   53 "ペルシアン"
   54 "コダック"
   55 "ゴルダック"
   56 "マンキー"
   57 "オコリザル"
   58 "ガーディ"
   59 "ウインディ"
   60 "ニョロモ"
   61 "ニョロゾ"
   62 "ニョロボン"
   63 "ケーシィ"
   64 "ユンゲラー"
   65 "フーディン"
   66 "ワンリキー"
   67 "ゴーリキー"
   68 "カイリキー"
   69 "マダツボミ"
   70 "ウツドン"
   71 "ウツボット"
   72 "メノクラゲ"
   73 "ドククラゲ"
   74 "イシツブテ"
   75 "ゴローン"
   76 "ゴローニャ"
   77 "ポニータ"
   78 "ギャロップ"
   79 "ヤドン"
   80 "ヤドラン"
   81 "コイル"
   82 "レアコイル"
   83 "カモネギ"
   84 "ドードー"
   85 "ドードリオ"
   86 "パウワウ"
   87 "ジュゴン"
   88 "ベトベター"
   89 "ベトベトン"
   90 "シェルダー"
   91 "パルシェン"
   92 "ゴース"
   93 "ゴースト"
   94 "ゲンガー"
   95 "イワーク"
   96 "スリープ"
   97 "スリーパー"
   98 "クラブ"
   99 "キングラー"
   100 "ビリリダマ"
   101 "マルマイン"
   102 "タマタマ"
   103 "ナッシー"
   104 "カラカラ"
   105 "ガラガラ"
   106 "サワムラー"
   107 "エビワラー"
   108 "ベロリンガ"
   109 "ドガース"
   110 "マタドガス"
   111 "サイホーン"
   112 "サイドン"
   113 "ラッキー"
   114 "モンジャラ"
   115 "ガルーラ"
   116 "タッツー"
   117 "シードラ"
   118 "トサキント"
   119 "アズマオウ"
   120 "ヒトデマン"
   121 "スターミー"
   122 "バリヤード"
   123 "ストライク"
   124 "ルージュラ"
   125 "エレブー"
   126 "ブーバー"
   127 "カイロス"
   128 "ケンタロス"
   129 "コイキング"
   130 "ギャラドス"
   131 "ラプラス"
   132 "メタモン"
   133 "イーブイ"
   134 "シャワーズ"
   135 "サンダース"
   136 "ブースター"
   137 "ポリゴン"
   138 "オムナイト"
   139 "オムスター"
   140 "カブト"
   141 "カブトプス"
   142 "プテラ"
   143 "カビゴン"
   144 "フリーザー"
   145 "サンダー"
   146 "ファイヤー"
   147 "ミニリュウ"
   148 "ハクリュー"
   149 "カイリュー"
   150 "ミュウツー"
   151 "ミュウ"
})

(def evolves
  [10  ;; キャタピー
   13  ;; ビードル
   16  ;; ポッポ
   19  ;; コラッタ
   21  ;; オニスズメ
   23  ;; アーボ
   25  ;; ピカチュウ
   35  ;; ピッピ
   37  ;; ロコン
   39  ;; プリン
   41  ;; ズバット
   43  ;; ナゾノクサ
   46  ;; パラス
   48  ;; コンパン
   50  ;; ディグダ
   52  ;; ニャース
   56  ;; マンキー
   54  ;; コダック
   69  ;; マダツボミ
   72  ;; メノクラゲ
   74  ;; イシツブテ
   77  ;; ポニータ
   79  ;; ヤドン
   81  ;; コイル
   86  ;; パウワウ
   88  ;; ベトベター
   90  ;; シェルダー
   96  ;; スリープ
   98  ;; クラブ
   100 ;; ビリリダマ
   102 ;; タマタマ
   104 ;; カラカラ
   109 ;; ドガース
   116 ;; タッツー
   118 ;; トサキント
   138 ;; オムナイト
   140 ;; カブト
   ])

(defprotocol IPokemonId
  (pokemon-id [_]))

(extend-protocol IPokemonId
  PokemonIdOuterClass$PokemonId
  (pokemon-id [pk]
    (let [descriptor (.getValueDescriptor pk)]
      {:name   (.getName descriptor)
       :number (.getNumber descriptor)
       :jname  (*pokemons* (.getNumber descriptor))}))
  Pokemon
  (pokemon-id [pk]
    (pokemon-id (.getPokemonId pk)))
  CatchablePokemon
  (pokemon-id [pk]
    (pokemon-id (.getPokemonId pk))))

(defprotocol PokemonStatus
  (status [_]))

(extend-protocol PokemonStatus
  Pokemon
  (status [pk]
    {:pokemon-id (pokemon-id pk)
     :cp (.getCp pk)
     :cp-multiplier (.getCpMultiplier pk)
     :individual-attack (.getIndividualAttack pk)
     :individual-defense (.getIndividualDefense pk)
     :individual-stamina (.getIndividualStamina pk)}))

(def catch-result-status
  {0 "を捕まえようとしたけどエラー！(捕まえようとしたボールが足りない？)"
   1 "を捕まえた！"
   2 "がボールから出た！"
   3 "が逃げ出した！"
   4 "にボールが当たらなかった！"})

(defprotocol CatchResultStatus
  (result-format [_]))

(extend-protocol CatchResultStatus
  CatchPokemonResponseOuterClass$CatchPokemonResponse$CatchStatus
  (result-format [_]
    (let [descriptor (.getValueDescriptor _)]
      (catch-result-status (.getNumber descriptor))))
  CatchResult
  (result-format [_]
    (-> _ .getStatus result-format)))
